File: ProblemSet/949. 斐波那契数列 II/949. 斐波那契数列 II.java
public class Solution {
    /**
     * @param n: an integer
     * @return: return an int
     */
    public int lastFourDigitsOfFn(int n) {
        if (n <= 1) return n;
        int[][] ans = new int[][]{{1, 0}, {0, 1}};
        int[][] A = new int[][]{{1, 1}, {1, 0}};
        --n;
        while (n > 0) {
            if ((n & 1) == 1) ans = matrixMul(ans, A);
            A = matrixMul(A, A);
            n >>= 1;
        }
        return ans[0][0];
    }

    private int[][] matrixMul(int[][] A, int[][] B) {
        int[][] tmp = new int[][]{{0, 0}, {0, 0}};
        for (int i = 0; i < 2; ++i) {
            for (int j = 0; j < 2; ++j) {
                for (int k = 0; k < 2; ++k)
                    tmp[i][j] = (tmp[i][j] + A[i][k] * B[k][j]) % 10000;
            }
        }
        return tmp;
    }
}

File: ProblemSet/949. 斐波那契数列 II/949. 斐波那契数列 II.py
class Solution:
    """
    @param n: an integer
    @return: return an int
    """
    def lastFourDigitsOfFn(self, n):
        if n <= 1:
            return n
        ans = [[1, 0], [0, 1]]
        a = [[1, 1], [1, 0]]
        n -= 1
        while n > 0:
            if n & 1:
                ans = self.matrix_mul(ans, a)
            a = self.matrix_mul(a, a)
            n >>= 1
        return ans[0][0]
        
    def matrix_mul(self, a, b):
        tmp = [[0, 0], [0, 0]]
        for i in range(2):
            for j in range(2):
                for k in range(2):
                    tmp[i][j] = (tmp[i][j] + a[i][k] * b[k][j]) % 10000

        return tmp